nav2_mppi_controller/src/collision_checker.cpp

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) 2025, Angsa Robotics GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#include "nav2_mppi_controller/collision_checker.hpp"
+#include "nav2_costmap_2d/inflation_layer.hpp"
+#include <algorithm>
+#include <cmath>
+
+namespace nav2_mppi_controller
+{
+
+MPPICollisionChecker::MPPICollisionChecker(
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros,
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node)
+: FootprintCollisionChecker(costmap_ros ? costmap_ros->getCostmap() : nullptr)
+{
+  if (node) {
+    clock_ = node->get_clock();
+    logger_ = node->get_logger();
+  }
+
+  if (costmap_ros) {
+    costmap_ros_ = costmap_ros;
+  }
+}
+
+void MPPICollisionChecker::setFootprint(
+  const nav2_costmap_2d::Footprint & footprint,
+  const bool & radius,
+  const std::string & inflation_layer_name)
+{
+  // Calculate the circumscribed cost automatically
+  possible_collision_cost_ = findCircumscribedCost(inflation_layer_name);
+
+  if (possible_collision_cost_ <= 0.0f) {
+    RCLCPP_ERROR_THROTTLE(
+      logger_, *clock_, 1000,
+      "Inflation layer either not found or inflation is not set sufficiently for "
+      "optimized non-circular collision checking capabilities. It is HIGHLY recommended to set"
+      " the inflation radius to be at MINIMUM half of the robot's largest cross-section. See "
+      "github.com/ros-planning/navigation2/tree/main/nav2_mppi_controller#potential-fields"
+      " for full instructions. This will substantially impact run-time performance.");
+  }
+
+  footprint_is_radius_ = radius;
+
+  // Use radius, no caching required
+  if (radius) {
+    return;
+  }
+
+  // Store the unoriented footprint for on-the-fly transformation
```

Comment:
We could consider bin-quantization like with SMAC if the tradeoff between tradeoff and accuracy is acceptable